### ISSUE TITLE
react/no-direct-mutation-state doesn't catch all cases

### ISSUE THREAD

**Issue by yantakus** [yannickcr/eslint-plugin-react#1418]
```javascript
  const { model } = this.state;
  model.rating = rate;
```

In this case eslint won't show any errors.

**Comment by jseminck** [comment]
I think we figured this out yesterday. Somebody created a PR for not allowing mutation for `this.props` and added support for this case. Once that PR is finished we'll try to move all the logic over to `no-direct-mutation-state` rule as well.

https://github.com/yannickcr/eslint-plugin-react/pull/1416

**Comment by meldafert** [comment]
is there any news on this? or alternatives that i could use?

**Comment by kangax** [comment]
@meldafert I just found out the other day about this rule not catching all cases. Ended up marking state immutable via Flow instead, which is perfect for this job:

```js
// ...
type State = $ReadOnly<{| ... |}>
class Smth extends Component<Props, State> {
  // ...
}
```

**Comment by meldafert** [comment]
@kangax thank you, this is exactly what i needed. (turns out typescript [has this too](https://basarat.gitbooks.io/typescript/content/docs/types/readonly.html)).

**Comment by kangax** [comment]
@meldafert lint rule pending https://github.com/gajus/eslint-plugin-flowtype/pull/400

**Comment by maniator** [comment]
> @kangax thank you, this is exactly what i needed. (turns out typescript [has this too](https://basarat.gitbooks.io/typescript/content/docs/types/readonly.html)).

TypeScript already disallows changing props as they are marked as readonly through typescript typing already (the app will not build if you try to mutate props)

**Comment by meldafert** [comment]
> @meldafert TypeScript already disallows changing props as they are marked as readonly through typescript typing already (the app will not build if you try to mutate props)
True, but only for mutating this.props, but not for eg. `this.props.list.push()`, which is what i'm trying to catch.
(if i declare the prop as `list: Readonly<Item[]>`, it does prevent me from using `push()`)

**Comment by maniator** [comment]
@meldafert ahh, so it does not do a _deap_ check.

You can probably also do (in ts)

```typescript
readonly list: Item[],
